[extension/awslogs_encoding] Add support for VPC flow logs in plain text file format

constanca-m opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

Add support for VPC flow logs sent to S3 in plain text format.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #38896.

<!--Describe what testing was performed and which tests were added.-->
#### Testing

There are new unit tests added.

<!--Describe the documentation added.-->
#### Documentation

Comments in the code and unit tests should be enough.

Fields are mapped this way:
| Flow log field               | Attribute                          | Available?     | Currently supported? |
|------------------------------|------------------------------------|----------------|----------------------|
| `version`                    | `aws.vpc.flow.log.version`         | :red_circle:   |                      |
| `account-id`                 | `cloud.account.id`                 | :green_circle: |                      |
| `interface-id`               | `aws.eni.id`                       | :red_circle:   |                      |
| `srcaddr`                    | `source.address`  or `network.peer.address`           | :red_circle:   |                      |
| `pkt-srcaddr`                | `source.address`                   | :green_circle: |                      |
| `dstaddr`                    | `destination.address`  or `network.peer.address`      | :red_circle:   |                      |
| `pkt-dstaddr`                | `destination.address`              | :green_circle: |                      |
| `srcport`                    | `source.port`                      | :green_circle: |                      |
| `dstport`                    | `destination.port`                 | :green_circle: |                      |
| `protocol`                   | `network.protocol.name`            | :green_circle: |                      |
| `packets`                    | `aws.vpc.flow.packets`             | :red_circle:   |                      |
| `bytes`                      | `aws.vpc.flow.bytes`               | :red_circle:   |                      |
| `start`                      | `aws.vpc.flow.start`               | :red_circle:   |                      |
| `end`                        | log timestamp                      | N/A            |                      |
| `action`                     | `aws.vpc.flow.action`              | :red_circle:   |                      |
| `log-status`                 | `aws.vpc.flow.status`              | :red_circle:   |                      |
| `vpc-id`                     | `aws.vpc.id`                       | :red_circle:   |                      |
| `subnet-id`                  | `aws.vpc.subnet.id`                | :red_circle:   |                      |
| `instance-id`                | `host.id`                          | :red_circle:   |                      |
| `tcp-flags`                  | `network.tcp.flags`                | :red_circle:   |                      |
| `type`                       | `network.type`                     | :green_circle: |                      |
| `region`                     | `cloud.region`                     | :green_circle: |                      |
| `az-id`                      | `aws.az.id`                        | :red_circle:   |                      |
| `sublocation-type`           | `aws.sublocation.type`             | :red_circle:   |                      |
| `sublocation-id`             | `aws.sublocation.id`               | :red_circle:   |                      |
| `pkt-src-aws-service`        | `aws.vpc.flow.source.service`      | :red_circle:   |                      |
| `pkt-dst-aws-service`        | `aws.vpc.flow.destination.service` | :red_circle:   |                      |
| `flow-direction`             | `network.io.direction`             | :green_circle: |                      |
| `traffic-path`               | `aws.vpc.flow.traffic_path`        | :red_circle:   |                      |
| `ecs-cluster-arn`            | `aws.ecs.cluster.arn`              | :green_circle: | :red_circle:         |
| `ecs-cluster-name`           | `aws.ecs.cluster.name`             | :red_circle:   | :red_circle:         |
| `ecs-container-instance-arn` | `aws.ecs.container.instance.arn`   | :red_circle:   | :red_circle:         |
| `ecs-container-instance-id`  | `aws.ecs.container.instance.id`    | :red_circle:   | :red_circle:         |
| `ecs-container-id`           | `aws.ecs.container.id`             | :red_circle:   | :red_circle:         |
| `ecs-second-container-id`    | `aws.ecs.second.container.arn`     | :red_circle:   | :red_circle:         |
| `ecs-service-name`           | `aws.ecs.service.name`             | :red_circle:   | :red_circle:         |
| `ecs-task-definition-arn`    | `aws.ecs.task.definition.arn`      | :red_circle:   | :red_circle:         |
| `ecs-task-arn`               | `aws.ecs.task.arn`                 | :green_circle: | :red_circle:         |
| `ecs-task-id`                | `aws.ecs.task.id`                  | :green_circle: | :red_circle:         |
| `reject-reason`              | `aws.vpc.flow.reject_reason`       | :red_circle:   |                      |

constanca-m committed:
Add support for VPC flow logs in plain text file format

constanca-m committed:
Add support for VPC flow logs in plain text file format

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+var (
+	supportedVPCFlowLogFileFormat = []string{fileFormatPlainText, fileFormatParquet}
+
+	supportedFieldValues = map[string][]string{
```
At least some of these should be translated to [SemConv attributes](https://opentelemetry.io/docs/specs/semconv/general/attributes/), but not all of them have a well defined attribute already. We'll probably need to raise an issue on https://github.com/open-telemetry/semantic-conventions and use provisional attribute names for now.

Looking at https://docs.aws.amazon.com/vpc/latest/userguide/flow-log-records.html#flow-logs-fields, I think we should be making these translations:

- `account-id` to [cloud.account.id](https://opentelemetry.io/docs/specs/semconv/attributes-registry/cloud/#cloud-account-id)
- `srcaddr` to [source.address](https://opentelemetry.io/docs/specs/semconv/attributes-registry/source/#source-address)
- `srcport` to [source.port](https://opentelemetry.io/docs/specs/semconv/attributes-registry/source/#source-port)
- `dstaddr` to [destination.address](https://opentelemetry.io/docs/specs/semconv/attributes-registry/destination/#destination-address)
- `dstport` to [destination.port](https://opentelemetry.io/docs/specs/semconv/attributes-registry/destination/#destination-port)
- `protocol` to [network.transport](https://opentelemetry.io/docs/specs/semconv/attributes-registry/network/#network-transport)
- `region` to [cloud.region](https://opentelemetry.io/docs/specs/semconv/attributes-registry/cloud/#cloud-region)

Ones I'm unsure about:
 - `az-id`: this is an availability zone ID not name. See https://docs.aws.amazon.com/ram/latest/userguide/working-with-az-ids.html for the distinction.
 - `pkt-srcaddr` and `pkg-dstaddr`. Not sure off hand how srcaddr/dstaddr/pkt-srcaddr/pkt-dstaddr should be translated, needs reading through SemConv in more detail.

I've left some other comments for other attribute translations. Some other fields should probably go under new `aws.*` attributes, e.g. `vpc-id`, `subnet-id`, `instance-id`, `sublocation-type`, `sublocation-id`. Some should go under a more specific `aws.vpc.flow` prefix, e.g. `action`, `packets`, `bytes`, `log-status`, `traffic-path`.

constanca-m replied:
Thanks! So the different between `pkt-srcaddr` and `srcaddr` is: `pkt-srcaddr` is always the original address, while `srcaddr` can be the address post NAT. For this reason, I picked `pkt-srcaddr` to be the `source.address`.

Since there are SO MANY fields, here is a table to make it easy to view without the code:

- The `Available?` column tells us if this is in the conventions or not.

| Flow log field               | Attribute                          | Available?     |
|------------------------------|------------------------------------|----------------|
| `version`                    | `aws.flow.log.version`             | :red_circle:   |
| `account-id`                 | `cloud.account.id`                 | :green_circle: |
| `interface-id`               | `aws.eni.id`                       | :red_circle:   |
| `srcaddr`                    | `source.layer.address`             | :red_circle:   |
| `pkt-srcaddr`                | `source.address`                   | :green_circle: |
| `dstaddr`                    | `destination.layer.address`        | :red_circle:   |
| `pkt-dstaddr`                | `destination.address`              | :green_circle: |
| `srcport`                    | `source.port`                      | :green_circle: |
| `dstport`                    | `destination.port`                 | :green_circle: |
| `protocol`                   | `network.protocol.number`          | :red_circle:   |
| `packets`                    | `aws.flow.log.packets`             | :red_circle:   |
| `bytes`                      | `aws.flow.log.bytes`               | :red_circle:   |
| `start`                      | `aws.flow.log.start`               | :red_circle:   |
| `end`                        | `aws.flow.log.end`                 | :red_circle:   |
| `action`                     | `aws.flow.log.action`              | :red_circle:   |
| `log-status`                 | `aws.flow.log.status`              | :red_circle:   |
| `vpc-id`                     | `aws.vpc.id`                       | :red_circle:   |
| `subnet-id`                  | `aws.subnet.id`                    | :red_circle:   |
| `instance-id`                | `aws.instance.id`                  | :red_circle:   |
| `tcp-flags`                  | `network.tcp.flags`                | :red_circle:   |
| `type`                       | `network.type`                     | :green_circle: |
| `region`                     | `cloud.region`                     | :green_circle: |
| `az-id`                      | `aws.az.id`                        | :red_circle:   |
| `sublocation-type`           | `aws.sublocation.type`             | :red_circle:   |
| `sublocation-id`             | `aws.sublocation.id`               | :red_circle:   |
| `pkt-src-aws-service`        | `aws.flow.log.source.service`      | :red_circle:   |
| `pkt-dst-aws-service`        | `aws.flow.log.destination.service` | :red_circle:   |
| `flow-direction`             | `network.io.direction`             | :green_circle: |
| `traffic-path`               | `aws.flow.log.traffic.path`        | :red_circle:   |
| `ecs-cluster-arn`            | `aws.ecs.cluster.arn`              | :green_circle: |
| `ecs-cluster-name`           | `aws.ecs.cluster.name`             | :red_circle:   |
| `ecs-container-instance-arn` | `aws.ecs.container.instance.arn`   | :red_circle:   |
| `ecs-container-instance-id`  | `aws.ecs.container.instance.id`    | :red_circle:   |
| `ecs-container-id`           | `aws.ecs.container.id`             | :red_circle:   |
| `ecs-second-container-id`    | `aws.ecs.second.container.arn`     | :red_circle:   |
| `ecs-service-name`           | `aws.ecs.service.name`             | :red_circle:   |
| `ecs-task-definition-arn`    | `aws.ecs.task.definition.arn`      | :red_circle:   |
| `ecs-task-arn`               | `aws.ecs.task.arn`                 | :green_circle: |
| `ecs-task-id`                | `aws.ecs.task.id`                  | :green_circle: |
| `reject-reason`              | `aws.flow.log.reject_reason`       | :red_circle:   |


What do you think of this? Any fields you would like to be refactored, or removed? Once we decide on it, we can open an issue to add the new attributes for the conventions.

This was so many fields, that maybe I made some mistake after looking at them for so long. Hopefully not.

axw replied:
Thank you, this is very helpful. My suggestions follow, to make progress over perfection:

- Let's handle ECS fields in a follow-up. Just drop them for now and we can handle them in a dedicated PR. We can take to take a look at what existing ECS-related components are doing for the fields you've marked as unavailable.
- For `srcaddr` and `dstaddr`: would it ever make sense that they're both different to `pkt-srcaddr` and `pkt-dstaddr` in the same record? IIANM it should only be one or the other, depending on the NAT direction -- so I think we could check for this and set [network.peer.address](https://opentelemetry.io/docs/specs/semconv/attributes-registry/network/#network-peer-address)
- These are somewhat more like metrics than logs, in that they represent an aggregation over a time window -- hence `start` and `end`. If we were to produce metrics, I think we would use `end` as the timestamp and `start` as the start timestamp. So long as we're representing them as logs, I think let's use `end` as the timestamp and store `start` as an attribute -- `aws.vpc.flow.start` is OK for now I guess, but I'm asking around to see if there's prior art.
- Instead of `network.protocol.number`, perhaps we could define a mapping from numbers to protocol keywords listed at https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml, and store as `network.transport`? i.e. 1 -> icmp, 6 -> tcp, etc.
- Perhaps `aws.vpc.subnet.id` instead of `aws.subnet.id`
- `aws.instance.id` should be [host.id](https://opentelemetry.io/docs/specs/semconv/attributes-registry/host/#host-id)
- I would suggest changing the "aws.flow" prefix to "aws.vpc.flow"

constanca-m replied:
> Instead of network.protocol.number, perhaps we could define a mapping from numbers to protocol keywords listed at https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml, and store as network.transport? i.e. 1 -> icmp, 6 -> tcp, etc.

I've thought of it. I couldn't find a public library that had this. Would you rather have this in this encoding?

axw replied:
I would prefer it. Maybe we can generate something from https://github.com/golang/net/tree/master/internal/iana?

constanca-m replied:
I have done some changes.

> Let's handle ECS fields in a follow-up. Just drop them for now and we can handle them in a dedicated PR. We can take to take a look at what existing ECS-related components are doing for the fields you've marked as unavailable.

Done! Added a `TODO` there.

> For srcaddr and dstaddr: would it ever make sense that they're both different to pkt-srcaddr and pkt-dstaddr in the same record? IIANM it should only be one or the other, depending on the NAT direction -- so I think we could check for this and set network.peer.address

Unfortunately, yes. See [example](https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-records-examples.html).

> These are somewhat more like metrics than logs, in that they represent an aggregation over a time window -- hence start and end. If we were to produce metrics, I think we would use end as the timestamp and start as the start timestamp. So long as we're representing them as logs, I think let's use end as the timestamp and store start as an attribute -- aws.vpc.flow.start is OK for now I guess, but I'm asking around to see if there's prior art.

Sorry, I got a bit confused here. Do you want to treat a flow log record as a metric?

> Perhaps aws.vpc.subnet.id instead of aws.subnet.id

Makes perfect sense. Changed it.

> aws.instance.id should be host.id

I had added as `aws.instance.id` as that refers to an EC2 instance (the [terraform name](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance) for that is also instance). I don't know enough about the field naming though, so I have changed to `host.id` now.

> I would suggest changing the aws.flow prefix to aws.vpc.flow. Maybe drop the "log" from "flow.log" - we're translating log records regarding a traffic flow, so the properties relate toe the flow rather than the log.

Fair enough. Changed it.

This is the new table (I'll be carrying this around before it makes its way to README):


| Flow log field               | Attribute                          | Available?     | Currently supported? |
|------------------------------|------------------------------------|----------------|----------------------|
| `version`                    | `aws.vpc.flow.version`             | :red_circle:   |                      |
| `account-id`                 | `cloud.account.id`                 | :green_circle: |                      |
| `interface-id`               | `aws.eni.id`                       | :red_circle:   |                      |
| `srcaddr`                    | `source.layer.address`             | :red_circle:   |                      |
| `pkt-srcaddr`                | `source.address`                   | :green_circle: |                      |
| `dstaddr`                    | `destination.layer.address`        | :red_circle:   |                      |
| `pkt-dstaddr`                | `destination.address`              | :green_circle: |                      |
| `srcport`                    | `source.port`                      | :green_circle: |                      |
| `dstport`                    | `destination.port`                 | :green_circle: |                      |
| `protocol`                   | `network.protocol.number`          | :red_circle:   |                      |
| `packets`                    | `aws.vpc.flow.packets`             | :red_circle:   |                      |
| `bytes`                      | `aws.vpc.flow.bytes`               | :red_circle:   |                      |
| `start`                      | `aws.vpc.flow.start`               | :red_circle:   |                      |
| `end`                        | `aws.vpc.flow.end`                 | :red_circle:   |                      |
| `action`                     | `aws.vpc.flow.action`              | :red_circle:   |                      |
| `log-status`                 | `aws.vpc.flow.status`              | :red_circle:   |                      |
| `vpc-id`                     | `aws.vpc.id`                       | :red_circle:   |                      |
| `subnet-id`                  | `aws.vpc.subnet.id`                | :red_circle:   |                      |
| `instance-id`                | `host.id`                          | :red_circle:   |                      |
| `tcp-flags`                  | `network.tcp.flags`                | :red_circle:   |                      |
| `type`                       | `network.type`                     | :green_circle: |                      |
| `region`                     | `cloud.region`                     | :green_circle: |                      |
| `az-id`                      | `aws.az.id`                        | :red_circle:   |                      |
| `sublocation-type`           | `aws.sublocation.type`             | :red_circle:   |                      |
| `sublocation-id`             | `aws.sublocation.id`               | :red_circle:   |                      |
| `pkt-src-aws-service`        | `aws.vpc.flow.source.service`      | :red_circle:   |                      |
| `pkt-dst-aws-service`        | `aws.vpc.flow.destination.service` | :red_circle:   |                      |
| `flow-direction`             | `network.io.direction`             | :green_circle: |                      |
| `traffic-path`               | `aws.vpc.flow.traffic.path`        | :red_circle:   |                      |
| `ecs-cluster-arn`            | `aws.ecs.cluster.arn`              | :green_circle: | :red_circle:         |
| `ecs-cluster-name`           | `aws.ecs.cluster.name`             | :red_circle:   | :red_circle:         |
| `ecs-container-instance-arn` | `aws.ecs.container.instance.arn`   | :red_circle:   | :red_circle:         |
| `ecs-container-instance-id`  | `aws.ecs.container.instance.id`    | :red_circle:   | :red_circle:         |
| `ecs-container-id`           | `aws.ecs.container.id`             | :red_circle:   | :red_circle:         |
| `ecs-second-container-id`    | `aws.ecs.second.container.arn`     | :red_circle:   | :red_circle:         |
| `ecs-service-name`           | `aws.ecs.service.name`             | :red_circle:   | :red_circle:         |
| `ecs-task-definition-arn`    | `aws.ecs.task.definition.arn`      | :red_circle:   | :red_circle:         |
| `ecs-task-arn`               | `aws.ecs.task.arn`                 | :green_circle: | :red_circle:         |
| `ecs-task-id`                | `aws.ecs.task.id`                  | :green_circle: | :red_circle:         |
| `reject-reason`              | `aws.vpc.flow.reject_reason`       | :red_circle:   |                      |

axw replied:
> For srcaddr and dstaddr: would it ever make sense that they're both different to pkt-srcaddr and pkt-dstaddr in the same record? IIANM it should only be one or the other, depending on the NAT direction -- so I think we could check for this and set network.peer.address
>
> Unfortunately, yes. See example.

I don't see an example of what this. To clarify, what I'm asking is: would it ever make sense to have a record where `srcaddr!=pkt-srcaddr && dstaddr!=pkt-dstaddr`?

There are definitely cases for:
 - `srcaddr==pkt-srcaddr && dstaddr!=pkt-dstaddr` (traffic out to internet through NAT gateway)
 - `srcaddr!=pkt-srcaddr && dstaddr==pkt-dstaddr` (return traffic from internet through NAT gateway)

But I don't think it would make sense to have both pairs different in the same record would it?

axw replied:
>> These are somewhat more like metrics than logs, in that they represent an aggregation over a time window -- hence start and end. If we were to produce metrics, I think we would use end as the timestamp and start as the start timestamp. So long as we're representing them as logs, I think let's use end as the timestamp and store start as an attribute -- aws.vpc.flow.start is OK for now I guess, but I'm asking around to see if there's prior art.
>
> Sorry, I got a bit confused here. Do you want to treat a flow log record as a metric?

I think it makes more sense tbh, but status quo is to store them as logs, and you can always convert the logs to metrics using [signaltometricsconnector](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/connector/signaltometricsconnector), so let's stick with this for now.

> aws.instance.id should be host.id
>
> I had added as aws.instance.id as that refers to an EC2 instance (the [terraform name](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance) for that is also instance). I don't know enough about the field naming though, so I have changed to host.id now.

I find it slightly awkward tbh, but this is the SemConv attribute that other components set for instance IDs.

constanca-m replied:
> I don't see an example of what this. To clarify, what I'm asking is: would it ever make sense to have a record where srcaddr!=pkt-srcaddr && dstaddr!=pkt-dstaddr?

Sorry, I misunderstood you.

I think it is possible. NAT and a load balancer, for example.

I settled for this logic: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ae4f140ad3aa651eb464f8c61398f9761b0ae7e8/extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go#L294-L322

WDYT?

constanca-m replied:
I fixed the logic now @axw . I have added a unit test with the same example you sent me from AWS

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+var (
+	supportedVPCFlowLogFileFormat = []string{fileFormatPlainText, fileFormatParquet}
+
+	supportedFieldValues = map[string][]string{
```
If/when we add support for ECS fields, we should follow https://opentelemetry.io/docs/specs/semconv/resource/cloud-provider/aws/ecs/

constanca-m replied:
Fields added. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38897#discussion_r2012241861. Closing this conversation in favor of that.

axw reviewed (changes requested):
Looks good overall, but I think we should be translating fields to semantic conventions, and considering additions to SemConv where necessary.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/testdata/valid_vpc_flow_log_expected.yaml:
```diff
@@ -0,0 +1,56 @@
+              - key: interface-id
+                value:
+                  stringValue: eni-0eb1e4178af74336c
+              - key: srcaddr
+                value:
+                  stringValue: '-'
```
I thought of this, but what if the user followed a custom format, and then we are just removing fields? Maybe it makes sense to leave as unknown, `-`, the same way AWS intended. WDYT?

constanca-m replied:
In [here](https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-s3-create-flow-log.html), it says:

>  For Log record format, specify the format for the flow log record.
> To use the default flow log record format, choose AWS default format.
> To create a custom format, choose Custom format. For Log format, choose the fields to include in the flow log record.

I'm thinking if we remove fields, the user will not see them, and maybe start debugging to see what is missing. When in reality it is just unknown.

axw replied:
We can add documentation that explains it. If users are customising their VPC flow logs, I expect they'll understand that some fields are sometimes going to unavailable. It's easier to add it later than it is to remove.

constanca-m replied:
Removed.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+		"tcp-flags":        {"1", "2", "4", "18"},
+		"type":             {"IPv4", "IPv6", "EFA"},
+		"sublocation-type": {"wavelength", "outpost", "localzone"},
+		"pkt-src-aws-service": {
+			"AMAZON", "AMAZON_APPFLOW", "AMAZON_CONNECT", "API_GATEWAY", "CHIME_MEETINGS", "CHIME_VOICECONNECTOR",
+			"CLOUD9", "CLOUDFRONT", "CODEBUILD", "DYNAMODB", "EBS", "EC2", "EC2_INSTANCE_CONNECT", "GLOBALACCELERATOR",
+			"KINESIS_VIDEO_STREAMS", "ROUTE53", "ROUTE53_HEALTHCHECKS", "ROUTE53_HEALTHCHECKS_PUBLISHING",
+			"ROUTE53_RESOLVER", "S3", "WORKSPACES_GATEWAYS",
+		},
+		"pkt-dst-aws-service": {
+			"AMAZON", "AMAZON_APPFLOW", "AMAZON_CONNECT", "API_GATEWAY", "CHIME_MEETINGS", "CHIME_VOICECONNECTOR",
+			"CLOUD9", "CLOUDFRONT", "CODEBUILD", "DYNAMODB", "EBS", "EC2", "EC2_INSTANCE_CONNECT", "GLOBALACCELERATOR",
+			"KINESIS_VIDEO_STREAMS", "ROUTE53", "ROUTE53_HEALTHCHECKS", "ROUTE53_HEALTHCHECKS_PUBLISHING",
+			"ROUTE53_RESOLVER", "S3", "WORKSPACES_GATEWAYS",
+		},
```
I expected this comment from you when I wrote this :laughing: How can I decide at which point it is too strict? Is it just this field, or all fields that have a check?

axw replied:
:)

IMO we should expect the logs are valid when they come in. These may be very high volume, so we should make sure the parsing code is not doing more than it needs to.

For some SemConv fields there are expectations about the values, e.g. `network.io.direction`. In that case it's not so much validation as it is special handling of each field for translation.

constanca-m replied:
I have removed the validate function.

constanca-m committed:
Address comments

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+		"action":           {"ACCEPT", "REJECT"},
+		"log-status":       {"OK", "NODATA", "SKIPDATA"},
+		"tcp-flags":        {"1", "2", "4", "18"},
+		"type":             {"IPv4", "IPv6", "EFA"},
```
Fields added. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38897#discussion_r2012241861. Closing this conversation in favor of that.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+			"KINESIS_VIDEO_STREAMS", "ROUTE53", "ROUTE53_HEALTHCHECKS", "ROUTE53_HEALTHCHECKS_PUBLISHING",
+			"ROUTE53_RESOLVER", "S3", "WORKSPACES_GATEWAYS",
+		},
+		"flow-direction": {"ingress", "egress"},
```
Fields added. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38897#discussion_r2012241861. Closing this conversation in favor of that.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+	supportedFieldValues = map[string][]string{
+		"action":           {"ACCEPT", "REJECT"},
+		"log-status":       {"OK", "NODATA", "SKIPDATA"},
+		"tcp-flags":        {"1", "2", "4", "18"},
```
Removed.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+		if supported == value {
+			return nil
+		}
+	}
```
Removed the validate function.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,196 @@
+
+func validateVPCFlowLog(field string, value string) error {
+	if value == "-" {
+		return nil
```
Thanks. I have changed it yesterday. Currently, we ignore these values.

constanca-m committed:
Change fields name, print warning

constanca-m committed:
add protocol

constanca-m committed:
make gotidy

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,332 @@
+		if err := addNumber(field, value, "aws.vpc.flow.version"); err != nil {
+			return false, err
+		}
+	case "srcaddr":
+		record.Attributes().PutStr("source.layer.address", value)
+	case "dstaddr":
+		record.Attributes().PutStr("destination.layer.address", value)
```
The default format only includes srcaddr and dstaddr, and not pkt-srcaddr or pkt-dstaddr. I think in that case we should still set the standard SemConv fields where we can. Also, as mentioned in the other thread, I think if all 4 are available then we should set https://opentelemetry.io/docs/specs/semconv/attributes-registry/network/#network-peer-address depending on the traffic direction.

Unfortunately the default format doesn't include traffic direction either, otherwise I think we could use that to pick the relevant address for `source.address` or `destination.address`, and the other for `network.peer.address`.

The only other option I can think of is: if pkt-srcaddr and pkt-dstaddr are not present, use srcaddr and dstaddr for `source.address` and `destination.address`. May be incorrect in the presence of NAT, but we can document this and how to resolve it (i.e. by including pkt-*addr).

constanca-m replied:
> The default format only includes srcaddr and dstaddr, and not pkt-srcaddr or pkt-dstaddr. 

We aren't settling for the default format. Even a custom format is included in the first line. That's what we are following, so I think we should handle this case already.

axw replied:
I understand we'll handle all the formats, what I meant is that a user may have configured a VPC flow log with just the default format, in which case pkt-* won't be present.

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,332 @@
+	case "region":
+		key.region = value
+	case "flow-direction":
+		record.Attributes().PutStr(conventions.AttributeNetworkIoDirection, value)
```
We should translate the value to the SemConv expected value here: egress -> transmit, ingress -> receive.

constanca-m replied:
Done.

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,332 @@
+		if err := addNumber(field, value, "aws.vpc.flow.start"); err != nil {
+			return false, err
+		}
+	case "end":
+		if err := addNumber(field, value, "aws.vpc.flow.end"); err != nil {
+			return false, err
+		}
```
This should be used to set the log record timestamp. I suggest dropping `aws.vpc.flow.end`, and just setting the timestamp.

constanca-m replied:
Done.

axw reviewed (commented):
Thanks for the updates, nearly there.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,332 @@
+}
+
+func (v *vpcFlowLogUnmarshaler) UnmarshalLogs(content []byte) (plog.Logs, error) {
+	var errGzipReader error
+	gzipReader, ok := v.gzipPool.Get().(*gzip.Reader)
+	if !ok {
+		gzipReader, errGzipReader = gzip.NewReader(bytes.NewReader(content))
+	} else {
+		errGzipReader = gzipReader.Reset(bytes.NewReader(content))
+	}
+	if errGzipReader != nil {
+		if gzipReader != nil {
+			v.gzipPool.Put(gzipReader)
+		}
+		return plog.Logs{}, fmt.Errorf("failed to decompress content: %w", errGzipReader)
+	}
+	defer func() {
+		_ = gzipReader.Close()
+		v.gzipPool.Put(gzipReader)
+	}()
```
I agree. Too many unmarshalers having this :)

constanca-m committed:
change names, adress rest of comments

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,332 @@
+	case "version":
+		if err := addNumber(field, value, "aws.vpc.flow.version"); err != nil {
+			return false, err
+		}
```
Done.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,332 @@
+	case "pkt-dst-aws-service":
+		record.Attributes().PutStr("aws.vpc.flow.destination.service", value)
+	case "traffic-path":
+		record.Attributes().PutStr("aws.vpc.flow.traffic.path", value)
```
Done.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/const.go:
```diff
@@ -0,0 +1,153 @@
+// protocolNames are needed to know the name of the protocol number given by the field
+// protocol in a flow log record.
+// See https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml.
+var protocolNames = map[int]string{
```
Done. Thanks for the suggestion!

constanca-m committed:
fix english

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/const.go:
```diff
@@ -0,0 +1,153 @@
+// protocolNames are needed to know the name of the protocol number given by the field
+// protocol in a flow log record.
+// See https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml.
+var protocolNames = [143]string{
```
I don't think the indices are all correct now. https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml says "nsh" is 145, but we only have 143 entries.

What I was suggesting before is to use this syntax:

```go
var protocolNames = [...]string{
  0: "hopopt",
  // etc.
  255: "reserved",
}
```

constanca-m replied:
Oh, you are absolutely correct :thinking:

constanca-m replied:
But then the easier way would be to have a map[int]string, don't you think? Otherwise we need empty values for numbers that do not map to any protocol name.

axw replied:
I think this is a fair tradeoff when there's only 256 entries.

constanca-m replied:
I have added empty metrics. I used a unit test to compare that everything was still correct (not commited). If the protocol is ` ` (empty), then the attribute protocol name will be empty. I decided to still have the protocol attribute for those cases. What do you think?

axw replied:
That's fine, but I think the syntax I used above is a little safer. You don't have to specify a value for every index that way - the unspecified ones will get an empty string. See https://go.dev/play/p/RKYse-UDfM6

constanca-m replied:
I was unaware this existed. Thanks!!

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,365 @@
+		if err := addNumber(field, value, "aws.vpc.flow.log.version"); err != nil {
+			return false, err
+		}
+	case "srcaddr":
+		// If pkt-srcaddr exists, then this might be an intermediary address.
+		// Otherwise, place it in source.address.
+		_, found := record.Attributes().Get(conventions.AttributeSourceAddress)
+		if found {
+			record.Attributes().PutStr("source.layer.address", value)
+		} else {
+			record.Attributes().PutStr(conventions.AttributeSourceAddress, value)
+		}
+	case "pkt-srcaddr":
+		// contrary to srcaddr logic
+		if intermediary, found := record.Attributes().Get(conventions.AttributeSourceAddress); found {
+			record.Attributes().PutStr("source.layer.address", intermediary.Str())
+		}
+		record.Attributes().PutStr(conventions.AttributeSourceAddress, value)
+	case "dstaddr":
+		// same as srcaddr
+		_, found := record.Attributes().Get(conventions.AttributeDestinationAddress)
+		if found {
+			record.Attributes().PutStr("destination.layer.address", value)
+		} else {
+			record.Attributes().PutStr(conventions.AttributeDestinationAddress, value)
+		}
+	case "pkt-dstaddr":
+		// same logic as pkt-srcaddr
+		if intermediary, found := record.Attributes().Get(conventions.AttributeSourceAddress); found {
+			record.Attributes().PutStr("destination.layer.address", intermediary.Str())
+		}
+		record.Attributes().PutStr(conventions.AttributeDestinationAddress, value)
```
Might be simpler to collect these particular fields and set the attributes after processing all fields.

axw replied:
Otherwise I _think_ the logic is OK, though I'm having trouble following because of all the if/elses and my brain is a bit fried.

I'm not a fan of the `foo.layer.address` names. I'd like to get to a conclusion on the other thread about whether there's ever a case where the two pairs are both different in one record.

constanca-m replied:
Resolving. Issue in discussion at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38897#discussion_r2015979155.

constanca-m committed:
fix address convention, fix protocol list

constanca-m committed:
Add benchmarking tests

constanca-m committed:
Merge branch 'main' into text-plain-vpc-flow-log

constanca-m committed:
Merge branch 'main' into text-plain-vpc-flow-log

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,396 @@
+	srcaddr, foundSrc := addresses["srcaddr"]
+	pktSrcaddr, foundSrcPkt := addresses["pkt-srcaddr"]
+	if !foundSrcPkt && foundSrc {
+		// there is no middle layer, and "srcaddr" field
+		// corresponds to the original source address.
```
```suggestion
		// there is no middle layer, assume "srcaddr" field
		// corresponds to the original source address.
```

It might not be the original source address, but this is the best we can do with the information we have

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,396 @@
+	dstaddr, foundDst := addresses["dstaddr"]
+	pktDstaddr, foundDstPkt := addresses["pkt-dstaddr"]
+	if !foundDstPkt && foundDst {
+		// there is no middle layer, and "dstaddr" field
+		// corresponds to the original destination address.
```
```suggestion
		// there is no middle layer, assume "dstaddr" field
		// corresponds to the original destination address.
```

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,396 @@
+	} else if foundDstPkt && foundDst {
+		recordAttr[conventions.AttributeDestinationAddress] = pktDstaddr
+		if pktDstaddr != dstaddr {
+			// dstaddr is the middle layer
```
Perhaps we should add a check here to see if `pktSrcaddr != srcaddr`, and log a warning in that case, as we will be discarding some data. Then we'll know if our assumption that only one of them differs is faulty.

constanca-m replied:
I have added a condition to see if we are overwritting the data and print a warning in that case.

axw replied:
Perfect, thank you!

axw reviewed (approved):
LGTM, just a few nits. Thanks for your persistence! :sweat_smile:

constanca-m committed:
Update extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go

Co-authored-by: Andrew Wilkins <axwalk@gmail.com>

constanca-m committed:
Update extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go

Co-authored-by: Andrew Wilkins <axwalk@gmail.com>

constanca-m committed:
print warning

constanca-m commented:
I add a benchmark test yesterday as well. All good there @axw ? It benchmarks `unmarshalPlainTextLogs` and not `UnmarshalLogs`, because the gzip took a lot of the statistics to itself.

constanca-m committed:
fix protocolNames

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/benchmark_test.go:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	for name, test := range tests {
+		data := createVPCFlowLogContent(b, filename, test.nLogs)
+		gzipReader, errGzipReader := gzip.NewReader(bytes.NewReader(data))
+		require.NoError(b, errGzipReader)
+
+		b.Run(name, func(b *testing.B) {
+			b.ReportAllocs()
+			for i := 0; i < b.N; i++ {
+				_, err := u.unmarshalPlainTextLogs(gzipReader)
+				require.NoError(b, err)
+			}
+		})
```
There's a bug here: the reader is depleted on the first iteration, and every subsequent iteration does nothing.

I think we should remove decompression from the benchmark. How about we update `unmarshalPlainTextLogs` to accept an `io.Reader` rather than `*gzip.Reader`, and skip compressing the input in the test.

constanca-m replied:
Thanks! That explains why once I changed to this test struct, the results got so much better: there was nothing else on the reader :laughing:

axw reviewed (changes requested):
@constanca-m I'm glad you asked about the benchmark. It looked fine, but I just ran it locally and found the time of each of the sizes was the same, which is surprising - they should be 10x as expensive at each level.

constanca-m committed:
fix gzip bug

axw reviewed (approved):
LGTM!

I think the 80MB test is probably unnecessary. It might be useful to have 8K and 8MB in case there's any constant per-file overhead, which we should see amortised by larger files. We can drop the 80MB one later though, it's fine.

constanca-m commented:
Alright, I don't mind dropping the 80MB file now, if you prefer? Which quantity of logs would fit the benchmark best? @axw

axw commented:
@constanca-m perhaps just two extremes: 1, and 1000. With 1 we'll see the maximum per-file overhead , and with 1000 it should be drowned out.

constanca-m committed:
change the number of logs on the benchmark

constanca-m committed:
Merge branch 'main' into text-plain-vpc-flow-log

constanca-m committed:
use strings.index

constanca-m commented:
@axw I have removed the `strings.split` like you mentioned, and replaced with `strings.Index`. It worked much better:

```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
                                             │ current.txt │            use-index.txt            │
                                             │   sec/op    │   sec/op     vs base                │
UnmarshalUnmarshalPlainTextLogs/1_log-16       2.265µ ± 2%   2.195µ ± 1%   -3.09% (p=0.000 n=10)
UnmarshalUnmarshalPlainTextLogs/1000_logs-16   660.7µ ± 1%   585.4µ ± 0%  -11.40% (p=0.000 n=10)
geomean                                        38.68µ        35.84µ        -7.34%

                                             │ current.txt  │            use-index.txt             │
                                             │     B/op     │     B/op      vs base                │
UnmarshalUnmarshalPlainTextLogs/1_log-16       5.617Ki ± 0%   5.398Ki ± 0%   -3.89% (p=0.000 n=10)
UnmarshalUnmarshalPlainTextLogs/1000_logs-16   756.4Ki ± 0%   537.6Ki ± 0%  -28.92% (p=0.000 n=10)
geomean                                        65.18Ki        53.87Ki       -17.35%

                                             │ current.txt │           use-index.txt            │
                                             │  allocs/op  │  allocs/op   vs base               │
UnmarshalUnmarshalPlainTextLogs/1_log-16        28.00 ± 0%    27.00 ± 0%  -3.57% (p=0.000 n=10)
UnmarshalUnmarshalPlainTextLogs/1000_logs-16   11.03k ± 0%   10.03k ± 0%  -9.07% (p=0.000 n=10)
geomean                                         555.7         520.3       -6.36%
```

Thanks for another great suggestion!

constanca-m committed:
Merge remote-tracking branch 'origin/text-plain-vpc-flow-log' into text-plain-vpc-flow-log

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,407 @@
+	// TODO Replace with an iterator starting from go 1.24:
+	// https://pkg.go.dev/strings#FieldsSeq
+	nFields := len(fields)
+	nValues := strings.Count(logLine, " ") + 1
+	if nFields != nValues {
+		return fmt.Errorf("expect %d fields per log line, got log line with %d fields", nFields, nValues)
+	}
```
For a followup:

Unless there's a bug in the parser, the number of fields should always match, so let's optimise for that. I suggest we remove this check here and count the fields as we're parsing to reduce overhead. Then at the very end we can return an error if the number of fields parsed does not match expectations.

constanca-m replied:
For sure, here is the issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39043.

axw commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go:
```diff
@@ -0,0 +1,407 @@
+	// See https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-records-examples.html#flow-log-example-nat.
+	ips := make(map[string]string, 4)
+
+	start := 0
+	for _, field := range fields {
+		var value string
+		end := strings.Index(logLine[start:], " ")
+		if end == -1 {
+			value = logLine[start:]
+		} else {
+			value = logLine[start : start+end]
+			start += end + 1 // skip the space
+		}
```
https://pkg.go.dev/strings#Cut should simplify this, something like:

```go
for _, field := range fields {
	if logLine == "" {
		// handle field count mismatch
	}
	var value string
	value, logLine, _ = strings.Cut(logLine, " ")
}
```

Let's do that in a followup though.

constanca-m replied:
Issue so we don't forget: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39043.

axw reviewed (approved):
@constanca-m there are several other performance improvements we can make, let's get this merged as it is and defer any more changes to followups please.
